### ISSUE TITLE
Remove "clinic" (aka families) from Contact filter dropdown, ideally remove need for `include_docs=true`

### ISSUE THREAD

**Issue by SCdF** [medic/medic-webapp#2218]
See #2209, #2210 and #2211.

Removing the families from the contact dropdown should improve both rendering time and bandwidth sent over the wire.

You could also look into working out why we need `include_docs=true` in the request and see if instead we can change the view to output what we need to render (which should be smaller than the whole doc). I'm pretty sure this view is only used for searching, so changing what it outputs shouldn't negatively affect anything else.

**Comment by abbyad** [comment]
That makes sense since it is the lowest level of the hierarchy and represented in the list. It will also make the contact filter considerably more usable.

**Comment by SCdF** [comment]
Hey @garethbowen when you're around would be good to chat about this code: https://github.com/medic/medic-webapp/blob/develop/static/js/controllers/inbox.js#L283-L376

Here is how I read it:
 - Whenever the page loads or a contact is CUD'd (it's like CRUD but without the Read you see…)
 - Load all [`PLACE_TYPES`](https://github.com/medic/medic-webapp/blob/develop/static/js/app.js#L104) (including the raw documents) and use that to generate a facility hierarchy (the thing that we care about it generating).
 - Load all [`CONTACT_TYPES`](https://github.com/medic/medic-webapp/blob/develop/static/js/app.js#L103) (including the raw documents) and set that to `$scope.people`
 - It's not then clear where $scope.people gets used (in the generation of the contact list maybe?)
 - It also appears to be used to populate some dropdowns somewhere, which sorts and filters those contacts _in memory, in Javascript_.

This seems a bit weird.

I suppose if you think of it from a pouch perspective it's all client side so it's not too relevant, but if I'm reading that right it seems like a really inefficient situation for admins.n

**Comment by SCdF** [comment]
Additionally, I'm pretty sure [this code](https://github.com/medic/medic-webapp/blob/develop/static/js/services/facility.js#L34-L38) implies that we also download everything and then filter it client side.

Let's talk after the call tonight @garethbowen, I'm still new enough to this code base to not be convinced I know what I'm talking about, or to be sure it's not some way for a good reason (and if I change it I'll end up subtly breaking something I didn't even know existed).

**Comment by SCdF** [comment]
[facility.js](https://github.com/medic/medic-webapp/blob/develop/static/js/services/facility.js) is in general a bit troubled I think. For posterity:

![2016-04-26 14 29 56](https://cloud.githubusercontent.com/assets/583851/14819850/ec9d75a6-0bbb-11e6-80cf-052c37683862.jpg)

Facility loads all contacts into memory regardless of what you ask for, including the actual docs. Worth checking to see if the callers of the various services listed (and their callers etc) actually care about the whole docs, or if they can actually get everything they want out a better way. It's also weird that facilities loads things that are people, which can't really be construed as facilities in any traditional sense.

Contact seems to be broken: it loads all contacts despite the clever feint where it pretends to care about the user's district. We should be able to remove that call / association entirely, unless there is subtle expectations that the various fail states of [UserDistrict](https://github.com/medic/medic-webapp/blob/develop/static/js/services/user.js#L26-L53) are checked here.

Most importantly though, it doesn't seem to be actually used anywhere. I'll triple check, but I think I can just delete it. Which would be great, because it removes the need for Facility to know about persons, and restricts it to much sensibler district hospital / health center / clinic set.

District should be using Facility but for some reason isn't. I think I can point it at Facility with no harm.

ChildFacility is weird, and only used by some stock analytics. I'm convinced it's [reduce](https://github.com/medic/medic-webapp/blob/develop/packages/kujua-reporting/kujua-reporting/views.js#L62-L64) function actually works, but I presume that when you call a view you're not actually invoking the reduce so maybe that's OK?

**Comment by SCdF** [comment]
FWIW I think it's also kind of weird that the [facilities](https://github.com/medic/medic-webapp/blob/develop/lib/views.js#L1-L22) view has people in it at all? I'm going to see if I can also change that as well, which will make the index smaller and faster.

**Comment by SCdF** [comment]
Removing `type='person'` from facilities takes the 42mb download every time down to 12mb (uncompressed, no idea what gzip will do to it). This is obviously much better, but we can still do better by working out if they need to include docs.

**Comment by SCdF** [comment]
Looks like the `Contact` service is still being used, it's just sneaky in how it's linked. It's used in the send-message module.

**Comment by SCdF** [comment]
Well if I knew that `push -f`ing to keep a branch rebased cleanly was going to spam tickets it referenced I would never have done it…

@garethbowen I think I've fixed this, and probably #2255, as well as taken a fairly good stab at addressing #2209, #2210, #2211 and #2258.

The relevant code changes are encompassed in this range: https://github.com/medic/medic-webapp/compare/0b677ef4cd4376a44921f6b50e3a024f96520669...08339a5fa192a6bacde5c8fbbd9f8bd9f6a812b5

If you could it a review that would be swell.

tl;dr; this replaces the Facilities implementation that caches all documents returned from the facilities view with one that does so individually by document type. This means that:
 - We only download what we (in theory) care about
 - Invalidation happens less frequently (editing a health_center won't invalidate all people)

This means that for normal use you should never download all person documents unless you foolishly click the send message button.

**In practice** this changes the code from:
 - A 3-4mb in PROD and ~50mb in dev download that occurred when the app loaded and reoccured whenever any Contact anywhere was edited
 - to a ??kb in PROD and a few hundred kb in dev download that occurs when the app is loaded and partially reoccurs depending on what you edit, though it usually doesn't, and if it does it's only 100kb or so.

If you look at individual commits (which are themselves squashed and editorialised from many other commits) you'll see I went on a bit of a journey before arriving back at "split the cache up". Not sure if you want to look at that detail though.

**Comment by garethbowen** [comment]
Yeah that looks great. Good thinking on caching per type. The only downside is multiple requests for the initial load, but fixing this would be quite complex I think, so happy with the solution unless we find it's a performance hit in production.
